### ISSUE TITLE
`findProxy` and `findProxyFromEnvironment` should be async

### ISSUE THREAD

**Issue by yamsergey** [dart-lang/sdk#44971]
Dart provides us with possibility to add proxy for HTTP/HTTPS requests. 

To do that, developers advised to implement one of the methods:

E.g. 
https://github.com/dart-lang/sdk/blob/d63d5d93a59b370bf5f5c277109110cb67a96b75/sdk/lib/_http/http.dart#L1776
https://github.com/dart-lang/sdk/blob/d63d5d93a59b370bf5f5c277109110cb67a96b75/sdk/lib/_http/http.dart#L1829

Also in documentation:
https://api.dart.dev/stable/2.10.5/dart-io/HttpClient/findProxy.html
https://api.flutter.dev/flutter/dart-io/HttpClient/findProxyFromEnvironment.html

Unfortunately all the options are synchronous. And at the same time  proxy configuration could be way more complicated than just `host` and `port` stored in environment variables. There could be `excluded urls` or `Auto Proxy configuration`, which will require async calls.

**Comment by marceloadsj** [comment]
@ChristianNavolskyi maybe this could help you:
https://github.com/BrowserStackCE/flutter_system_proxy/blob/main/lib/flutter_system_proxy.dart#L73

It's for flutter but maybe you can find something useful over there.

**Comment by ChristianNavolskyi** [comment]
Thanks for the suggestion. I tried it out, but somehow it always returned `"undefined"` for the URLs I passed it instead of `"DIRECT"` or the actual proxy.

During debugging, I saw that it got the pac file correctly, but it seems like it has problems evaluating it on the native side.

Do you have any idea why that could be?

With another package I found on GitLab it works, so the pac file and everything should be correct.

**Comment by marceloadsj** [comment]
Hummm, are you able to post your pac file here? (Ofc, feel free to hide any sensitive info, IPs or anything!)

I just want to double check it to see if I find anything weird. Hehe

**Comment by ChristianNavolskyi** [comment]
Yes of course, had to change the file ending to txt because GitHub doesn't support pac. 
[proxy-default.txt](https://github.com/dart-lang/sdk/files/10244773/proxy-default.txt)

Thanks for your support!

**Comment by marceloadsj** [comment]
Alright, the .pac seems to be okay. I will try to test the setup soon and let you know.

For my other app, I used the "advanced setup" from https://github.com/BrowserStackCE/flutter_system_proxy#advanced-usage-custom-dio-adapter

Let me know if you find something. I do the same from my side. Hehe 😅
